core/vm/contracts_write.go

```diff
@@ -150,3 +163,163 @@ func (c *stakingPrecompile) RunWriteCapable(
+	}
+}
+
+type crossShardXferPrecompile struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *crossShardXferPrecompile) RequiredGas(
+	evm *EVM,
+	contract *Contract,
+	input []byte,
+) (uint64, error) {
+	// We charge gas here to make sure contracts using `delegatecall`
+	// are not able to subsidize the base gas for cross shard transfers by EOAs
+	var payload []byte = make([]byte, 0)
+	fromAddress, toAddress, fromShardID, toShardID, value, err :=
+		parseCrossShardXferData(evm, contract, input)
+	if err == nil {
+		data := struct {
+			fromAddress common.Address
+			toAddress   common.Address
+			fromShardID uint32
+			toShardID   uint32
+			value       *big.Int
+		}{
+			fromAddress: fromAddress,
+			toAddress:   toAddress,
+			fromShardID: fromShardID,
+			toShardID:   toShardID,
+			value:       value,
+		}
+		if encoded, err := rlp.EncodeToBytes(data); err == nil {
+			payload = encoded
+		}
+	}
+	if gas, err := IntrinsicGas(
+		payload,
+		false,                                   // contractCreation
+		evm.ChainConfig().IsS3(evm.EpochNumber), // homestead
+		evm.ChainConfig().IsIstanbul(evm.EpochNumber), // istanbul
+		false, // isValidatorCreation
+	); err != nil {
+		return 0, err // ErrOutOfGas occurs when gas payable > uint64
+	} else {
+		return gas, nil
+	}
+}
+
+// RunWriteCapable runs the actual contract
+func (c *crossShardXferPrecompile) RunWriteCapable(
+	evm *EVM,
+	contract *Contract,
+	input []byte,
+) ([]byte, error) {
+	fromAddress, toAddress, fromShardID, toShardID, value, err :=
+		parseCrossShardXferData(evm, contract, input)
+	if err != nil {
+		return nil, err
+	}
+	// validate not a contract (toAddress can still be a contract)
+	if len(evm.StateDB.GetCode(fromAddress)) > 0 && !evm.IsValidator(evm.StateDB, fromAddress) {
+		return nil, errors.New("cross shard xfer not yet implemented for contracts")
+	}
+	// can't have too many shards
+	if toShardID >= evm.Context.NumShards {
+		return nil, errors.New("toShardId out of bounds")
+	}
+	// not for simple transfers
+	if fromShardID == toShardID {
+		return nil, errors.New("from and to shard id can't be equal")
+	}
+	// make sure nobody sends extra or less money
+	if contract.Value().Cmp(value) != 0 {
+		return nil, errors.New("argument value and msg.value not equal")
+	}
+	// now do the actual transfer
+	// step 1 -> remove funds from the precompile address
+	evm.Transfer(evm.StateDB, contract.Address(), toAddress, value, types.SubtractionOnly)
+	// step 2 -> make a cross link
+	// note that the transaction hash is added by state_processor.go to this receipt
+	// and that the receiving shard does not care about the `From` but we use the original
+	// instead of the precompile address for consistency
+	evm.CXReceipt = &types.CXReceipt{
```

Comment:
what happens if the code calls the precompile multiple times in the same transaction? We probably need a list of CXReceipt, right? Or is there anyway to prevent repeated call to the precompile in the same txn?

Reply:
> Or is there anyway to prevent repeated call to the precompile in the same txn?

Aren't we already doing that by checking that the `From` address is not a contract?